File: supabase/functions/push-due-reminders/winter_tyre_period_test.ts
import { assertEquals } from 'jsr:@std/assert@1'
import {
  nextSeasonalSwap,
  swapsSeasonally,
  winterTyrePeriodFor,
} from './winter_tyre_period.ts'

const day = (y: number, m: number, d: number) => new Date(Date.UTC(y, m - 1, d))

Deno.test('Croatia is dated and ignores the weather', () => {
  assertEquals(winterTyrePeriodFor('HR').rule, 'dated')
  assertEquals(winterTyrePeriodFor('HR').start, [11, 15])
})

Deno.test('an unverified country claims nothing', () => {
  for (const code of ['GB', 'US', 'IT', 'ZZ', '']) {
    assertEquals(winterTyrePeriodFor(code).rule, 'none')
  }
})

Deno.test('autumn points at the winter date', () => {
  const swap = nextSeasonalSwap('HR', day(2026, 10, 1))!
  assertEquals(swap.date.getTime(), day(2026, 11, 15).getTime())
  assertEquals(swap.direction, 'to_winter')
})

Deno.test('the first day of the window is the swap, not a missed one', () => {
  const swap = nextSeasonalSwap('HR', day(2026, 11, 15))!
  assertEquals(swap.date.getTime(), day(2026, 11, 15).getTime())
})

Deno.test('the day after points at the spring swap', () => {
  const swap = nextSeasonalSwap('HR', day(2026, 11, 16))!
  assertEquals(swap.date.getTime(), day(2027, 4, 15).getTime())
  assertEquals(swap.direction, 'to_summer')
})

Deno.test('midwinter crosses the year end', () => {
  const swap = nextSeasonalSwap('HR', day(2026, 12, 31))!
  assertEquals(swap.date.getTime(), day(2027, 4, 15).getTime())
})

Deno.test('Slovenia comes out of winter a month before Croatia', () => {
  const swap = nextSeasonalSwap('SI', day(2026, 12, 1))!
  assertEquals(swap.date.getTime(), day(2027, 3, 15).getTime())
})

Deno.test('a country with no window has no swap to point at', () => {
  for (const code of ['DE', 'GB', 'US', 'IT']) {
    assertEquals(nextSeasonalSwap(code, day(2026, 10, 1)), null)
  }
})

Deno.test('all-season tyres mean no seasonal swap', () => {
  assertEquals(
    swapsSeasonally([{ season: 'all_season', retired_at: null }]),
    false,
  )
})

Deno.test('a winter set means there is a swap', () => {
  assertEquals(swapsSeasonally([{ season: 'winter', retired_at: null }]), true)
})

Deno.test('no tyres recorded is not evidence of all-season', () => {
  assertEquals(swapsSeasonally([]), true)
})

Deno.test('a retired seasonal set does not keep the swap alive', () => {
  assertEquals(
    swapsSeasonally([
      { season: 'winter', retired_at: '2025-01-01' },
      { season: 'all_season', retired_at: null },
    ]),
    false,
  )
})

File: test/domain/maintenance/winter_tyre_period_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:garage/domain/maintenance/winter_tyre_period.dart';

DateTime day(int year, int month, int dayOfMonth) =>
    DateTime.utc(year, month, dayOfMonth);

void main() {
  group('what a country requires', () {
    test('Croatia is dated and does not care about the weather', () {
      final period = winterTyrePeriodFor('HR');

      expect(period.rule, WinterTyreRule.dated);
      expect(period.start, const MonthDay(11, 15));
      expect(period.end, const MonthDay(4, 15));
    });

    test('Austria is dated but only binds when the road is wintry', () {
      final period = winterTyrePeriodFor('AT');

      expect(period.rule, WinterTyreRule.datedWhenWintry);
      expect(period.start, const MonthDay(11, 1));
      expect(period.end, const MonthDay(4, 15));
    });

    test('Germany has no dates at all', () {
      final period = winterTyrePeriodFor('DE');

      expect(period.rule, WinterTyreRule.whenWintryOnly);
      expect(period.start, isNull);
      expect(period.end, isNull);
    });

    test('a country the app has not verified claims nothing', () {
      for (final code in ['GB', 'US', 'IT', 'ZZ', '']) {
        expect(
          winterTyrePeriodFor(code).rule,
          WinterTyreRule.none,
          reason: '$code should not be given a window nobody checked',
        );
      }
    });

    test('the country code is read case-insensitively', () {
      expect(winterTyrePeriodFor('hr').rule, WinterTyreRule.dated);
    });
  });

  group('the next swap in Croatia', () {
    SeasonalSwap? next(DateTime today) =>
        nextSeasonalSwap(countryCode: 'HR', today: today);

    test('autumn points at the winter date', () {
      final swap = next(day(2026, 10, 1))!;

      expect(swap.date, day(2026, 11, 15));
      expect(swap.direction, SwapDirection.toWinter);
    });

    test('the first day of the window is the swap, not a missed one', () {
      // The tyres have to be on *by* the 15th, so on the 15th the swap is due
      // today rather than six months away.
      final swap = next(day(2026, 11, 15))!;

      expect(swap.date, day(2026, 11, 15));
      expect(swap.direction, SwapDirection.toWinter);
    });

    test('the day after, the next thing to do is the spring swap', () {
      final swap = next(day(2026, 11, 16))!;

      expect(swap.date, day(2027, 4, 15));
      expect(swap.direction, SwapDirection.toSummer);
    });

    test('midwinter crosses the year end into spring', () {
      final swap = next(day(2026, 12, 31))!;

      expect(swap.date, day(2027, 4, 15));
      expect(swap.direction, SwapDirection.toSummer);
    });

    test('after the spring date the next is the following winter', () {
      final swap = next(day(2026, 4, 16))!;

      expect(swap.date, day(2026, 11, 15));
      expect(swap.direction, SwapDirection.toWinter);
    });
  });

  test('Slovenia comes out of winter a month before Croatia', () {
    final swap = nextSeasonalSwap(countryCode: 'SI', today: day(2026, 12, 1))!;

    expect(swap.date, day(2027, 3, 15));
    expect(swap.direction, SwapDirection.toSummer);
  });

  test('a country with no window has no swap to point at', () {
    for (final code in ['DE', 'GB', 'US', 'IT']) {
      expect(
        nextSeasonalSwap(countryCode: code, today: day(2026, 10, 1)),
        isNull,
        reason: '$code has no statutory date to pin a reminder to',
      );
    }
  });
}
